Foundation/NSURLSession/MultiHandle.swift

```diff
@@ -99,25 +103,33 @@ fileprivate extension URLSession._MultiHandle {
         // through libdispatch (DispatchSource) and store the source(s) inside
         // a `SocketSources` which we in turn store inside libcurl's multi handle
         // by means of curl_multi_assign() -- we retain the object fist.
-        let action = _SocketRegisterAction(rawValue: CFURLSessionPoll(value: what))
-        var socketSources = _SocketSources.from(socketSourcePtr: socketSourcePtr)
-        if socketSources == nil && action.needsSource {
-            let s = _SocketSources()
-            let p = Unmanaged.passRetained(s).toOpaque()
-            CFURLSessionMultiHandleAssign(rawHandle, socket, UnsafeMutableRawPointer(p))
-            socketSources = s
-        } else if socketSources != nil && action == .unregister {
-            // We need to release the stored pointer:
-            if let opaque = socketSourcePtr {
-                Unmanaged<_SocketSources>.fromOpaque(opaque).release()
+        commandQueue.async {
+            self.cancelInProgress?.wait()
+            self.cancelInProgress = nil
```

Comment:
The nil assignment is guarded by the `wait()` and only the cancel handler will `signal()` - which mean the nil assignment is executed only when the cancel handler no longer needs the semaphore. The assignment to `nil` signifies that there is no cancel operation in progress. Increasing the scope of the semaphore will lead to other `register` commands (may be on unrelated sockets) to `wait` with nobody to `signal` them.